circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/copier/CopierContext.java

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (C) 2016-2020 Expedia, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.api.copier;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.hadoop.fs.Path;
+
+import com.hotels.bdp.circustrain.api.conf.TableReplication;
+
+public class CopierContext {
```

Comment:
thoughts on making this immutable? Is there a risk of a copier accidentally change one of these fields which might affect further copiers in the chain?

Reply:
Yeah, I thought about making this immutable and having a builder instead of this big constructor but this was quicker and easier to do in terms of updating the existing code. However the TableReplication isn't immutable so we'd only be halfway there unless I looked into that too (not sure that's really possible as it maps directly to the config which needs to set values).

If we didn't want to use a builder I could add 2 more constructors which allow the TableReplication to be passed in too and then I make immutable copies of the List and Map member variables and we live with TableReplication not being immutable.

I can do either (or if someone has better suggestions) if we think it's worth it?

Reply:
I think since this is a new one we make the good example and make it immutable. I'm ok leaving TableReplication for another time as that will be backward incompatible change I guess.

Reply:
Isn't there any place where the context may actually need to be updated?

Reply:
yeah I was also thinking about that, not sure what we want to achieve. The order in which copiers are run is guaranteed by spring so you could have a copier relying on another copier updating the context. If that's they way we want to go then obviously we don't want to make it immutable. Could argue either way I guess :)

Reply:
At the moment I don't see any of the copiers updating the context but yes, you could argue that that could be a useful feature ;) My suggestion would be to follow our usual best practice and make them immutable for now, if we ever find we need some way of modifying/sharing context between copiers we could break that and make them mutable (or at that point perhaps some other solution presents itself).

Reply:
If it keeps on being backward-compatible and not too difficult to do, go ahead with making it immutable then. All the other things being equal, immutable objects are always preferable.

Reply:
I've changed it to be as immutable as I think I can :)

Reply:
Looks good!

Reply:
the class needs to be final

Reply:
It does! Good spot!

Reply:
and because TableReplication isn't immutable you need to copy it as well. Path also isn't :/ it's a pandemic. Not sure how much you want to open that can of worms. I can live with a not fully immutable class for the sake of pragmatism.

Reply:
Yeah, you're right, I think we either leave as it is now and live with a decent _attempt_ but `TableReplication` and `Path` are mutable or we just roll the whole set of immutable changes back.

Reply:
I'm ok with this.